### PR TITLE
Fix queen spider AI breaking on vomit egg

### PR DESCRIPTION
**Pull request by CheffieGithub**
## About the PR 
fix issue where queen clown spiders get stuck on vomiting egg

## Why's this needed? 
bad bug

### CODE REVIEW

File: code/mob/living/critter/ai/spider.dm [hunk 1]
```diff
@@ -21,11 +21,42 @@
...
+/datum/aiTask/sequence/goalbased/critter/vomit_egg/New(parentHolder, transTask)
+	..(parentHolder, transTask)
+	add_task(holder.get_instance(/datum/aiTask/succeedable/critter/vomit_egg, list(holder)))
+
+/datum/aiTask/sequence/goalbased/critter/vomit_egg/precondition()
+	var/mob/living/critter/spider/clownqueen/C = holder.owner
+	var/datum/targetable/critter/vomitegg/egg = C.abilityHolder.getAbility(/datum/targetable/critter/vomitegg)
+	if(!egg)
+		egg = C.abilityHolder.getAbility(/datum/targetable/critter/vomitegg/cluwne)
+	return !egg.disabled && (length(C.babies) < max_spiders)
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/ai/spider.dm:42]: Clownspiders don't remove themselves from this list on death, which is a GC issue and will mean this makes them only ever spawn 4 total.

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]

File: code/mob/living/critter/ai/spider.dm (hunk 1 above)
Comment (CheffieGithub) [reply code/mob/living/critter/ai/spider.dm:42]: yeah I found this out while I was testing it but I dunno how to fix

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
If you're going to make them weakrefs then you'll need to update the other code that accesses that list (it's just the spider `was_harmed` proc I think) to deref them properly, also probably best stick a comment on it to say it's full of weakrefs.

**Comment by CheffieGithub** [comment]
forgot about that part

File: code/mob/living/critter/spider.dm [hunk 2]
```diff
@@ -495,7 +495,10 @@
 			return
 		var/defenders = 0		//this is the amount of babies that will defend you
 		var/count = 0
-		for (var/mob/living/critter/spider/clown/CS in babies)
+		for (var/datum/weakref/ref as anything in babies)
+			var/mob/living/critter/spider/clown/CS = ref.deref()
+			if (CS == null)
+				babies.Remove(ref)
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/spider.dm:501]: you'll want to `continue` here or it's probably still going to runtime below (might get caught by `GET_DIST`, but I wouldn't rely on it)

File: code/mob/living/critter/spider.dm [hunk 3]
```diff
@@ -495,7 +495,10 @@
 			return
 		var/defenders = 0		//this is the amount of babies that will defend you
 		var/count = 0
-		for (var/mob/living/critter/spider/clown/CS in babies)
+		for (var/datum/weakref/ref as anything in babies)
+			var/mob/living/critter/spider/clown/CS = ref.deref()
+			if (CS == null)
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/spider.dm:500]:
```suggestion
			if (!istype(CS))
```

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]

**Pull request merged**